File: Team2/Vani Chitkara/Spiral Matrix.cpp
Problem: https://leetcode.com/problems/spiral-matrix/

Solution:

vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int n=matrix.size();
        if(n==0)
            return {};
		vector<int> arr;
        int m=matrix[0].size();
        if(n==1){
            for(int i=0;i<m;i++){
                arr.push_back(matrix[0][i]);                
            }
            return arr;
        }
        if(m==1){
            for(int i=0;i<n;i++){
                arr.push_back(matrix[i][0]);                
            }
            return arr;
        }
        int top=0,left=0,right=m-1,bottom=n-1;
        while(top<=bottom && left<=right){
            for(int i=left;i<=right;i++){
                arr.push_back(matrix[top][i]);
            }
            top++;
            for(int i=top;i<=bottom;i++){
                arr.push_back(matrix[i][right]);
            }
            right--;
            if(top<=bottom){
                for(int i=right;i>=left;i--){
                    arr.push_back(matrix[bottom][i]);
                }
                bottom--;
            }
            if(left<=right){
                for(int i=bottom;i>=top;i--)
                    arr.push_back(matrix[i][left]);
                left++;
            }
        }
        return arr;
    }

File: Team1/Riya Tanwar/spiral-matrix.java

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        
        List<Integer> spiralOrderList = new ArrayList<>();
        
        if(matrix.length == 0) return spiralOrderList;

        int top = 0, bottom = matrix.length - 1; 
        int left = 0, right = matrix[0].length - 1; 

        while(top <= bottom && left <= right) {
            for(int i = left; i <= right; i++)
                spiralOrderList.add(matrix[top][i]);
            
            top++;
            
            for(int i = top; i <= bottom; i++)
                spiralOrderList.add(matrix[i][right]);
            
            right--;
            
            for(int i = right; i >= left; i--) {

                if(top > bottom) 
                    break;

                spiralOrderList.add(matrix[bottom][i]);

            }

            bottom--;

            for(int i = bottom; i >= top; i--) {
                
                if(left > right) 
                    break;

                spiralOrderList.add(matrix[i][left]);
            }

            left++;

        }

        return spiralOrderList;
    }
}
